test-utils/src/main/scala/com/mesosphere/utils/mesos/MesosTest.scala

```diff
@@ -0,0 +1,394 @@
+                                         |  "fault_domain":
+                                         |    {
+                                         |      "region":
+                                         |        {
+                                         |          "name": "${faultDomain.region}"
+                                         |        },
+                                         |      "zone":
+                                         |        {
+                                         |          "name": "${faultDomain.zone}"
+                                         |        }
+                                         |    }
+                                         |}
+        """.stripMargin
+
+      val nodeAttributes = Map(
+        "fault_domain_region" -> Some(faultDomain.region.value),
+        "fault_domain_zone" -> Some(faultDomain.zone.value))
+      (nodeAttributes, Some(mesosFaultDomainCmdOption))
+    } else Map.empty -> None
+
+    // uniquely identify each agent node, useful for constraint matching
+    val attributes: Map[String, Option[String]] = Map("node" -> Some(i.toString)) ++ faultDomainAgentAttributes
+
+    val renderedAttributes: String = attributes.map { case (key, maybeVal) => s"$key${maybeVal.map(v => s":$v").getOrElse("")}" }.mkString(";")
+
+    Agent(resources = new Resources(ports = PortAllocator.portsRange(), gpus = agentsGpus), extraArgs = Seq(
+      s"--attributes=$renderedAttributes"
+    ) ++ mesosFaultDomainAgentCmdOption.map(fd => s"--domain=$fd"))
+  }
+
+  if (autoStart) {
+    start()
+  }
+
+  def start(): Future[String] = async {
+    masters.foreach(_.start())
+    agents.foreach(_.start())
+    val masterUri = await(waitForLeader())
+    await(waitForAgents(masterUri))
+    masterUri
+  }
+
+  def waitForLeader(): Future[String] = async {
+    val firstMaster = s"http://${masters.head.ip}:${masters.head.port}"
+    val result = Retry("wait for leader", maxAttempts = Int.MaxValue, maxDuration = waitForMesosTimeout) {
+      Http().singleRequest(Get(firstMaster + "/redirect")).map { result =>
+        result.discardEntityBytes() // forget about the body
+        if (result.status.isFailure()) {
+          throw new Exception(s"Couldn't determine leader: $result")
+        }
+        result
+      }
+    }
+
+    def maybeFixURI(uri: String): String = {
+      // some versions of mesos issue redirects with broken Location headers; fix them here
+      if (uri.indexOf("//") == 0) {
+        "http:" + uri
+      } else {
+        uri
+      }
+    }
+
+    val location = await(result).headers.find(_.lowercaseName() == "location").map(_.value()).getOrElse(firstMaster)
+    maybeFixURI(location)
+  }
+
+  def waitForAgents(masterUri: String): Future[Done] = {
+    val mesosFacade = new MesosFacade(masterUri)
+    Retry.blocking("wait for agents", maxAttempts = Int.MaxValue, maxDuration = waitForMesosTimeout) {
+      val numAgents = mesosFacade.state.value.agents.size
```

Comment:
See. Here we use it ;). Also, we should just use `eventually` here.

Reply:
Bam!